src/python/pants/backend/docker/commands.py

```diff
@@ -0,0 +1,268 @@
+
+    def register(self, dockerfile_attrs: Dict[str, Any]) -> None:
+        dockerfile_attrs["entry_point"] = self
+
+    def encode_arg(self) -> Generator[str, None, None]:
+        if self.form is EntryPoint.Form.EXEC:
+            yield json.dumps([self.executable, *(self.arguments or [])])
+        else:
+            yield self.executable
+            if self.arguments:
+                yield " ".join(self.arguments)
+
+    @classmethod
+    def decode_arg(cls, arg: str) -> Dict[str, Any]:
+        if arg.startswith("["):
+            form = EntryPoint.Form.EXEC
+            cmd_line = json.loads(arg)
+        else:
+            form = EntryPoint.Form.SHELL
+            cmd_line = arg.split(" ")
+        return dict(executable=cmd_line[0], arguments=tuple(cmd_line[1:]), form=form)
+
+
+@dataclass(frozen=True)
+class Copy(DockerfileCommand):
+    """The COPY instruction copies new files or directories from <src> and adds them to the
+    filesystem of the container at the path <dest>.
+
+        COPY [--chown=<user>:<group>] <src>... <dest>
+        COPY [--chown=<user>:<group>] ["<src>",... "<dest>"]
+
+        COPY --from=<name> ...
+
+    https://docs.docker.com/engine/reference/builder/#copy
+    """
+
+    _command = "COPY"
+
+    src: Tuple[str, ...]
+    dest: str
+    chown: Optional[str] = None
+    copy_from: Optional[str] = None
+
+    _arg_regexp = re.compile(
+        r"""
+        ^
+        # optional flags
+        (?:
+          (--chown=(?P<chown>\S+)\s+)
+          |
+          (--from=(?P<copy_from>\S+)\s+)
+        )*
+
+        # paths, will be post processed
+        (?P<paths>.+)
+        $
+        """,
+        re.VERBOSE,
+    )
+
+    def register(self, dockerfile_attrs: Dict[str, Any]) -> None:
+        self._append("copy", dockerfile_attrs)
+
+    def encode_arg(self) -> Generator[str, None, None]:
+        if self.copy_from:
+            yield f"--from={self.copy_from}"
+
+        if self.chown:
+            yield f"--chown={self.chown}"
+
+        paths = [*self.src, self.dest]
+        if any(" " in s for s in paths if s):
+            yield json.dumps(paths)
+        else:
+            for path in paths:
+                yield path
+
+    @classmethod
+    def decode_arg(cls, arg: str) -> Dict[str, Optional[str]]:
+        args: Dict[str, Any] = cls._decode_arg_regexp(cls._arg_regexp, arg)
+        paths_string = args.pop("paths") or ""
+        if paths_string.startswith("["):
+            paths = json.loads(paths_string)
+        else:
+            paths = paths_string.split()
+
+        if len(paths) > 1:
+            args["src"] = tuple(paths[:-1])
+            args["dest"] = paths[-1]
+        return args
+
+    # "RUN": ,
+    # "CMD": ,
+    # "LABEL": ,
+    # "EXPOSE": ,
+    # "ENV": ,
+    # "ADD": ,
+    # "COPY": ,
+    # "VOLUME": ,
+    # "USER": ,
+    # "WORKDIR": ,
+    # "ARG": ,
+    # "ONBUILD": ,
+    # "STOPSIGNAL": ,
+    # "HEALTHCHECK": ,
+    # "SHELL": ,
```

Comment:
Should probably clean up these comment lines before merging.. All defined valid Dockerfile commands.